README.rst
Kinto Changes
=============

.. image:: https://github.com/Kinto/kinto-changes/actions/workflows/test.yml/badge.svg
        :target: https://github.com/Kinto/kinto-changes/actions

.. image:: https://img.shields.io/pypi/v/kinto-changes.svg
        :target: https://pypi.python.org/pypi/kinto-changes

.. image:: https://coveralls.io/repos/Kinto/kinto-changes/badge.svg?branch=master
        :target: https://coveralls.io/r/Kinto/kinto-changes

**kinto-changes** shows the list of collection timestamps, allowing to poll changes
on several collections with one HTTP request.


Install
-------

::

    pip install kinto-changes

Setup
-----

In the `Kinto <http://kinto.readthedocs.io/>`_ settings:

.. code-block :: ini

    kinto.includes = kinto_changes

    # List of buckets/collections to show:
    kinto.changes.resources = /buckets/settings
                              /buckets/blocklists/collections/certificates


Cache Control
'''''''''''''

Like `cache control in Kinto collections <https://kinto.readthedocs.io/en/stable/api/1.x/collections.html#collection-caching>`_, it is possible to configure ``Cache-Control`` headers via some settings:

.. code-block:: ini

    kinto.monitor.changes.record_cache_expires_seconds = 60

If cache busting query parameters then responses can be cached more agressively.
If the setting below is set then a different cache control expiration will be set:

.. code-block:: ini

    kinto.monitor.changes.record_cache_maximum_expires_seconds = 3600


Advanced options
''''''''''''''''

**host**

The changes entries will have a ``host`` attribute, that can be used to
distinguish changes from several Kinto instances.

.. code-block :: ini

    kinto.changes.http_host = website.domain.tld

By default, it will rely on the global setting ``kinto.http_host``.


**_since sanetizing**

When reaching the monitor/changes collection, if the provided ``_since`` query parameter
is too old, we redirect the clients to the full list of changes (ie. without ``_since``).

Set this setting to control the maximum age allowed. Set to ``-1`` to disable redirection.

.. code-block :: ini

    kinto.changes.since_max_age_days = 21

Redirects are cached by default. Set to ``-1`` to disable, or ``0`` to cache forever.

.. code-block :: ini

    kinto.changes.since_max_age_redirect_ttl_seconds = 86400


Endpoints
---------

* ``GET /v1/buckets/monitor/collections/changes/records``.

Returns the list of timestamps.

* ``GET /v1/buckets/{bid}/collections/{cid}/changeset?_expected={timestamp}``.

Returns the following response for the collection:

- ``changes``: list of records, optionnally filtered with ``?_since="{timestamp}"``
- ``metadata``: collection attributes
- ``timestamp``: records timestamp

Note: the ``_expected={}`` querystring parameter is mandatory but can contain any
arbitrary value (used for cache busting).


CHANGELOG.rst
Changelog
=========

4.1.2 (unreleased)
------------------

- Nothing changed yet.


4.1.1 (2021-11-15)
------------------

**Bug fixes**

- Fix redirection for old ``?_since`` values on changeset endpoints (fixes #299)


4.1.0 (2021-10-28)
------------------

**New features**

- Send cache control headers on old ``?_since`` redirects (cache for 1 day by default).


4.0.0 (2021-10-21)
------------------

**Breaking Changes**

- No ``ResourceEvent`` is sent anymore for changes in this virtual monitor/changes collection.


3.2.0 (2020-11-20)
------------------

**New features**

- Add filtering support in monitor/changes changeset endpoint


3.1.2 (2020-09-30)
------------------

**Bug fixes**

- Fix crash when collection is unknown and backend running in readonly mode


3.1.1 (2020-09-28)
------------------

**Bug fixes**

- Bypass storage when redirecting on old ``_since``


3.1.0 (2020-09-22)
------------------

**New features**

- Redirect clients whose ``?_since`` value is too old (21 days by default)
- Add support for monitor/changes in changeset endpoints (fixes #173)

**Bug fixes**

- Fix validation rule for ``_since`` query parameter


3.0.2 (2020-06-02)
------------------

**Bug fixes**

- Include tombstones in changeset endpoint (fixes #175)


3.0.1 (2020-05-26)
------------------

- Sort changes explicitly by `last_modified` descending

3.0.0 (2020-04-20)
------------------

**Breaking changes**

- Switch changeset timestamp from ETag to integer (#168)


2.2.0 (2020-03-26)
------------------

**New features**

- Add changeset endpoint (#154)

**Bug fixes**

- Fix crash with ``If-None-Match: *`` (fixes #158)


2.1.1 (2019-12-23)
------------------

**Bug fixes**

- Use explicit resource timestamp from DB instead of event payload


2.1.0 (2019-02-20)
------------------

**New feature**

- Send ``Cache-Control`` headers if cache busting query parameters or concurrency control request headers are sent (#66)

2.0.0 (2019-01-15)
------------------

**Breaking changes**

- Requires Kinto >= 12


1.3.0 (2018-10-15)
------------------

**New feature**

- Add ability to configure cache control headers (#47)


1.2.1 (2018-09-17)
------------------

**Bug fixes**

- Don't do anything during a ``migrate`` command (fixes #43).


1.2.0 (2018-09-06)
------------------

**New feature**

- Events are now generated on the monitor/changes collection (#41).


1.1.1 (2018-04-17)
------------------

**Internal changes**

- Get rid of six


1.1.0 (2017-11-16)
------------------

**Breaking changes**

- Upgrade to Kinto 7


**Bug fixes**

- Disable reify to always get the most accurate timestamp. (#36)


1.0.0 (2017-02-14)
------------------

**Bug fixes**

- Accessing the monitoring collection when no changes occured don't fail anymore (fixes #23)
- The timestamps shown in the monitoring endpoint are now **exactly equal** (never superior anymore)
  to the timestamps of the monitored collections.

**Breaking changes**

* The change endpoint **location is now hard-coded** (``/buckets/monitor/collections/changes/records``)
  and cannot be configured.
* The permissions principals cannot be specified anymore.
  The change endpoint is now **always public**.
* The ``monitor`` bucket and ``changes`` collection are not required anymore and
  are not created anymore.
* ``POST`` and ``DELETE`` are not supported on the changes endpoint anymore.
* Individual entries (eg. ``/buckets/monitor/collections/changes/records/{id}``)
  cannot be accessed anymore.
* The listener was dropped. Configuration must be changed:

Before:

.. code-block :: ini

    kinto.event_listeners = changes
    kinto.event_listeners.changes.use = kinto_changes.listener
    kinto.event_listeners.changes.http_host = website.domain.tld
    kinto.event_listeners.changes.collections = /buckets/settings
                                                /buckets/blocklists/collections/certificates

Now:

.. code-block :: ini

    kinto.changes.http_host = website.domain.tld
    kinto.changes.resources = /buckets/settings
                              /buckets/blocklists/collections/certificates


0.5.0 (2017-01-16)
------------------

- Do not force the timestamp of monitored entries (#27)


0.4.0 (2016-11-07)
------------------

- Add the plugin version in the capability (#20)
- Add collections in the capability (#18)
- Add a specific setting to override global ``http_host`` value (#24)

0.3.0 (2016-05-19)
------------------

- Update to ``kinto.core`` for compatibility with Kinto 3.0. This
  release is no longer compatible with Kinto < 3.0, please upgrade!


0.2.0 (2016-04-25)
------------------

- Addition of the changes capability

0.1.0 (2015-12-22)
------------------

- Initial code.
- Bucket and collection name configuration.
- Changes read permissions configuration.
- Selection of buckets and collections to follow configuration.



tests/__init__.py
import os
import configparser

from kinto import main as kinto_main
from kinto.core.testing import get_user_headers, BaseWebTest as CoreWebTest


here = os.path.abspath(os.path.dirname(__file__))


class BaseWebTest(CoreWebTest):
    api_prefix = "v1"
    entry_point = kinto_main
    config = 'config.ini'

    def __init__(self, *args, **kwargs):
        super(BaseWebTest, self).__init__(*args, **kwargs)
        self.headers.update(get_user_headers('mat'))
        self.headers.update({'Origin': 'http://localhost:9999'})

    @classmethod
    def get_app_settings(cls, extras=None):
        ini_path = os.path.join(here, cls.config)
        config = configparser.ConfigParser()
        config.read(ini_path)
        settings = dict(config.items('app:main'))
        settings.update(extras or {})
        return settings

    def setUp(self):
        super(BaseWebTest, self).setUp()
        self.create_collection('blocklists', 'certificates')

    def create_collection(self, bucket_id, collection_id):
        bucket_uri = '/buckets/%s' % bucket_id
        self.app.put_json(bucket_uri,
                          {},
                          headers=self.headers)
        collection_uri = bucket_uri + '/collections/%s' % collection_id
        self.app.put_json(collection_uri,
                          {},
                          headers=self.headers)


tests/listener.py
def load_from_config():
    return ()


tests/test_utils.py
import unittest
from unittest import mock

from pyramid.request import Request

from kinto_changes.utils import changes_object


class ChangesRecordTest(unittest.TestCase):
    def test_single_hardcoded(self):
        request = Request.blank(path="/")
        request.route_path = mock.Mock()
        request.route_path.return_value = "/buckets/a/collections/b"
        request.registry = mock.Mock()
        request.registry.settings = {}
        timestamp = 1525457597166
        entry = changes_object(request, "a", "b", timestamp)

        self.assertEqual(
            entry,
            {
                "bucket": "a",
                "collection": "b",
                "host": "",
                "id": "9527d115-6191-fa49-a530-8fbfc4997755",
                "last_modified": timestamp,
            },
        )

    def test_another_hardcoded(self):
        request = Request.blank(path="/")
        request.route_path = mock.Mock()
        request.route_path.return_value = "/buckets/a/collections/b"
        request.registry = mock.Mock()
        request.registry.settings = {"http_host": "https://localhost:443"}
        timestamp = 1525457597166
        entry = changes_object(request, "a", "b", timestamp)

        self.assertEqual(
            entry,
            {
                "bucket": "a",
                "collection": "b",
                "host": "https://localhost:443",
                "id": "fa48a96d-1600-f561-8645-3395acb08a5a",
                "last_modified": timestamp,
            },
        )


kinto_changes/utils.py
import hashlib
from uuid import UUID

from pyramid.settings import aslist

from kinto.core import utils as core_utils


def monitored_collections(registry):
    storage = registry.storage
    resources_uri = aslist(registry.settings.get('changes.resources', ''))
    collections = []

    for resource_uri in resources_uri:
        resource_name, matchdict = core_utils.view_lookup_registry(registry, resource_uri)
        if resource_name == 'bucket':
            # Every collections in this bucket.
            result = storage.list_all(resource_name='collection', parent_id=resource_uri)
            collections.extend([(matchdict['id'], obj['id']) for obj in result])

        elif resource_name == 'collection':
            collections.append((matchdict['bucket_id'], matchdict['id']))

    return collections


def changes_object(request, bucket_id, collection_id, timestamp):
    """Generate an object for /buckets/monitor/collections/changes."""
    http_host = request.registry.settings.get('http_host') or ''
    collection_uri = core_utils.instance_uri(
        request, 'collection', bucket_id=bucket_id, id=collection_id)
    uniqueid = http_host + collection_uri
    identifier = hashlib.md5(uniqueid.encode('utf-8')).hexdigest()
    entry_id = str(UUID(identifier))

    return dict(id=entry_id,
                last_modified=timestamp,
                bucket=bucket_id,
                collection=collection_id,
                host=http_host)


kinto_changes/__init__.py
import pkg_resources
from pyramid.settings import aslist


#: Module version, as defined in PEP-0396.
__version__ = pkg_resources.get_distribution(__package__).version

MONITOR_BUCKET = "monitor"
MONITOR_BUCKET_PATH = "/buckets/{}".format(MONITOR_BUCKET)
CHANGES_COLLECTION = "changes"
CHANGES_COLLECTION_PATH = "{}/collections/{}".format(
    MONITOR_BUCKET_PATH, CHANGES_COLLECTION
)
CHANGES_RECORDS_PATH = "{}/records".format(CHANGES_COLLECTION_PATH)
CHANGESET_PATH = "/buckets/{bid}/collections/{cid}/changeset"


def includeme(config):
    settings = config.get_settings()
    collections = settings.get("changes.resources", [])

    config.add_api_capability(
        "changes",
        version=__version__,
        description="Track modifications of records in Kinto and store"
        " the collection timestamps into a specific bucket"
        " and collection.",
        url="http://kinto.readthedocs.io/en/latest/tutorials/"
        "synchronisation.html#polling-for-remote-changes",
        collections=aslist(collections),
    )

    config.scan("kinto_changes.views")


kinto_changes/listener.py
"""Listener that mirrors events from any collection into the changes collection."""

from kinto.core import utils as core_utils
from kinto.core.events import notify_resource_event, ACTIONS
from .utils import monitored_collections, changes_object, is_monitoring_collection


class Listener(object):
    def __init__(self, config):
        self.registry = config.registry
        self.collection_timestamps = {}

    def track_timestamps(self, event):
        for (bucket_id, collection_id) in monitored_collections(self.registry):
            timestamp = self.get_collection_timestamp(bucket_id, collection_id)
            self.collection_timestamps[(bucket_id, collection_id)] = timestamp

    def get_collection_timestamp(self, bucket_id, collection_id):
        collection_uri = core_utils.instance_uri_registry(self.registry,
                                                          'collection',
                                                          bucket_id=bucket_id,
                                                          id=collection_id)
        return self.registry.storage.resource_timestamp(parent_id=collection_uri,
                                                        resource_name='record')

    def on_record_changed(self, event):
        bucket_id = event.payload['bucket_id']
        collection_id = event.payload['collection_id']
        timestamp = self.get_collection_timestamp(bucket_id, collection_id)

        if not is_monitoring_collection(self.registry, bucket_id, collection_id):
            return

        for change in event.impacted_records:
            # This might be the first we've seen of this collection.
            old_timestamp = self.collection_timestamps.get((bucket_id, collection_id), None)

            # Synthesize event for /buckets/monitor/collections/changes record.
            new = changes_object(event.request, bucket_id, collection_id, timestamp)
            old = None
            action = ACTIONS.CREATE
            if old_timestamp:
                action = ACTIONS.UPDATE
                old = changes_object(event.request, bucket_id, collection_id, old_timestamp)

            resource_data = {'bucket_id': 'monitor', 'collection_id': 'changes', 'id': new['id']}

            notify_resource_event(
                event.request, '/buckets/monitor/collections/changes',
                timestamp, new, action, old=old,
                resource_name='record', resource_data=resource_data
            )

            self.collection_timestamps[(bucket_id, collection_id)] = timestamp


kinto_changes/views.py
from datetime import datetime, timedelta
from urllib.parse import urlencode

import colander
import kinto.core
from cornice.validators import colander_validator
from kinto.authorization import RouteFactory
from kinto.core import resource
from kinto.core import utils as core_utils
from kinto.core.storage import Filter, Sort
from kinto.core.storage import exceptions as storage_exceptions
from kinto.core.storage.memory import extract_object_set
from kinto.core.utils import COMPARISON, instance_uri
from pyramid import httpexceptions
from pyramid.security import IAuthorizationPolicy
from zope.interface import implementer

from . import (
    CHANGES_COLLECTION,
    CHANGES_COLLECTION_PATH,
    CHANGES_RECORDS_PATH,
    CHANGESET_PATH,
    MONITOR_BUCKET,
)
from .utils import changes_object, monitored_collections


class ChangesModel(object):
    id_field = "id"
    modified_field = "last_modified"
    deleted_field = "deleted"
    permissions_field = "__permissions__"

    def __init__(self, request):
        self.request = request
        self.storage = request.registry.storage

        self.__entries = None

    def timestamp(self):
        if not self._entries():
            return core_utils.msec_time()
        max_value = max([e["last_modified"] for e in self._entries()])
        return max_value

    def get_objects(
        self,
        filters=None,
        sorting=None,
        pagination_rules=None,
        limit=None,
        include_deleted=False,
        parent_id=None,
    ):
        objs, _ = extract_object_set(
            objects=self._entries(),
            filters=filters,
            sorting=sorting,
            pagination_rules=pagination_rules,
            limit=limit,
        )
        return objs

    def _entries(self):
        if self.__entries is None:
            self.__entries = {}

            for (bucket_id, collection_id) in monitored_collections(
                self.request.registry
            ):
                collection_uri = core_utils.instance_uri(
                    self.request, "collection", bucket_id=bucket_id, id=collection_id
                )
                timestamp = self.storage.resource_timestamp(
                    parent_id=collection_uri, resource_name="record"
                )
                entry = changes_object(
                    self.request, bucket_id, collection_id, timestamp
                )
                self.__entries[entry[self.id_field]] = entry

        return self.__entries.values()


class ChangesSchema(resource.ResourceSchema):
    host = colander.SchemaNode(colander.String())
    bucket = colander.SchemaNode(colander.String())
    collection = colander.SchemaNode(colander.String())

    class Options:
        preserve_unknown = False


@implementer(IAuthorizationPolicy)
class AnonymousRoute(RouteFactory):
    def check_permission(self, principals, bound_perms):
        # Bypass permissions check on /buckets/monitor.
        return True


@resource.register(
    name="changes",
    description="List of changes",
    plural_path=CHANGES_RECORDS_PATH,
    object_path=None,
    plural_methods=("GET",),
    factory=AnonymousRoute,
)
class Changes(resource.Resource):

    schema = ChangesSchema

    def __init__(self, request, context=None):
        # Bypass call to storage if _since is too old.
        _handle_old_since_redirect(request)
        # Inject custom model.
        self.model = ChangesModel(request)
        super(Changes, self).__init__(request, context)

    def plural_get(self):
        result = super().plural_get()
        _handle_cache_expires(self.request, MONITOR_BUCKET, CHANGES_COLLECTION)
        return result


def _handle_cache_expires(request, bid, cid):
    # If the client sends cache busting query parameters, then we can cache more
    # aggressively.
    settings = request.registry.settings
    prefix = f"{bid}.{cid}.record_cache"
    default_expires = settings.get(f"{prefix}_expires_seconds")
    maximum_expires = settings.get(f"{prefix}_maximum_expires_seconds", default_expires)

    has_cache_busting = "_expected" in request.GET
    cache_expires = maximum_expires if has_cache_busting else default_expires

    if cache_expires is not None:
        request.response.cache_expires(seconds=int(cache_expires))


def _handle_old_since_redirect(request):
    """
    In order to limit the number of possible combinations
    of `_since` and `_expected` querystring parameters,
    and thus maximize the effect of caching, we redirect the clients
    that arrive here with a very old `_since` value.

    This simply means that these clients will have to iterate
    and compare the local timestamps of the whole list of changes
    instead of a filtered subset.

    https://searchfox.org/mozilla-central/rev/b58ca450/services/settings/remote-settings.js#299

    See https://bugzilla.mozilla.org/show_bug.cgi?id=1529685
    and https://bugzilla.mozilla.org/show_bug.cgi?id=1665319#c2
    """
    try:
        # request.validated is not populated yet (resource was not instantiated yet,
        # we want to bypass storage).
        qs_since_str = request.GET.get("_since", "")
        qs_since = int(qs_since_str.strip('"'))
    except ValueError:
        # Will fail later during resource querystring validation.
        return

    settings = request.registry.settings
    max_age_since = int(settings.get("changes.since_max_age_days", 21))
    if max_age_since < 0:
        # Redirect is disabled.
        return

    min_since_dt = datetime.now() - timedelta(days=max_age_since)
    min_since = min_since_dt.timestamp() * 1000

    if qs_since >= min_since:
        # Since value is recent. No redirect.
        return

    http_scheme = settings.get("http_scheme") or "https"
    http_host = settings.get(
        "changes.http_host", request.registry.settings.get("http_host")
    )
    host_uri = f"{http_scheme}://{http_host}"
    redirect = host_uri + request.matched_route.generate(request.matchdict)

    queryparams = request.GET.copy()
    del queryparams["_since"]
    if queryparams:
        redirect += "?" + urlencode(queryparams)

    # Serve a redirection, with optional cache control headers.
    response = httpexceptions.HTTPTemporaryRedirect(redirect)
    cache_seconds = int(
        settings.get("changes.since_max_age_redirect_ttl_seconds", 60 * 60 * 24)
    )
    if cache_seconds >= 0:
        response.cache_expires(cache_seconds)
    raise response


@implementer(IAuthorizationPolicy)
class ChangeSetRoute(RouteFactory):
    """The changeset endpoint should have the same permissions as the collection
    metadata.

    The permission to read records is implicit when metadata are readable.
    """

    def __init__(self, request):
        super().__init__(request)
        bid = request.matchdict["bid"]
        cid = request.matchdict["cid"]
        collection_uri = instance_uri(request, "collection", bucket_id=bid, id=cid)
        # This route context will be the same as when reaching the collection URI.
        self.permission_object_id = collection_uri
        self.required_permission = "read"

    def check_permission(self, principals, bound_perms):
        # The monitor/changes changeset endpoint is publicly accesible.
        if self.permission_object_id == CHANGES_COLLECTION_PATH:
            return True
        # Otherwise rely on the collection permissions.
        return super().check_permission(principals, bound_perms)


changeset = kinto.core.Service(
    name="collection-changeset", path=CHANGESET_PATH, factory=ChangeSetRoute
)


class QuotedTimestamp(colander.SchemaNode):
    """Integer between "" used in _since querystring."""

    schema_type = colander.String
    error_message = "The value should be integer between double quotes."
    validator = colander.Regex('^"([0-9]+?)"$', msg=error_message)

    def deserialize(self, cstruct=colander.null):
        param = super(QuotedTimestamp, self).deserialize(cstruct)
        if param is colander.drop:
            return param
        return int(param.strip('"'))


class ChangeSetQuerystring(colander.MappingSchema):
    _since = QuotedTimestamp(missing=colander.drop)
    _expected = colander.SchemaNode(colander.String())
    _limit = colander.SchemaNode(colander.Integer(), missing=colander.drop)
    # Query parameters used on monitor/changes endpoint.
    bucket = colander.SchemaNode(colander.String(), missing=colander.drop)
    collection = colander.SchemaNode(colander.String(), missing=colander.drop)


class ChangeSetSchema(colander.MappingSchema):
    querystring = ChangeSetQuerystring()


@changeset.get(
    schema=ChangeSetSchema(), permission="read", validators=(colander_validator,)
)
def get_changeset(request):
    bid = request.matchdict["bid"]
    cid = request.matchdict["cid"]

    storage = request.registry.storage

    queryparams = request.validated["querystring"]
    limit = queryparams.get("_limit")
    filters = []
    include_deleted = False
    if "_since" in queryparams:
        filters = [Filter("last_modified", queryparams["_since"], COMPARISON.GT)]
        # Include tombstones when querying with _since
        include_deleted = True

    if (bid, cid) == (MONITOR_BUCKET, CHANGES_COLLECTION):
        # Redirect old since, on monitor/changes only.
        _handle_old_since_redirect(request)

        if "bucket" in queryparams:
            filters.append(Filter("bucket", queryparams["bucket"], COMPARISON.EQ))

        if "collection" in queryparams:
            filters.append(
                Filter("collection", queryparams["collection"], COMPARISON.EQ)
            )

        model = ChangesModel(request)
        metadata = {}
        timestamp = model.timestamp()
        changes = model.get_objects(
            filters=filters, limit=limit, include_deleted=include_deleted
        )

    else:
        bucket_uri = instance_uri(request, "bucket", id=bid)
        collection_uri = instance_uri(request, "collection", bucket_id=bid, id=cid)

        try:
            # We'll make sure that data isn't changed while we read metadata, changes, etc.
            before = storage.resource_timestamp(
                resource_name="record", parent_id=collection_uri
            )
            # Fetch collection metadata.
            metadata = storage.get(
                resource_name="collection", parent_id=bucket_uri, object_id=cid
            )

        except storage_exceptions.ObjectNotFoundError:
            raise httpexceptions.HTTPNotFound()

        except storage_exceptions.BackendError as e:
            # The call to `resource_timestamp()` on an empty collection will try initialize it.
            # If the instance is read-only, it fails with a backend error. Raise 404 in this
            # case otherwise raise the original backend error.
            if "when running in readonly" in str(e):
                raise httpexceptions.HTTPNotFound()
            raise

        # Fetch list of changes.
        changes = storage.list_all(
            resource_name="record",
            parent_id=collection_uri,
            filters=filters,
            limit=limit,
            id_field="id",
            modified_field="last_modified",
            deleted_field="deleted",
            sorting=[Sort("last_modified", -1)],
            include_deleted=include_deleted,
        )
        # Fetch current collection timestamp.
        timestamp = storage.resource_timestamp(
            resource_name="record", parent_id=collection_uri
        )

        # Do not serve inconsistent data.
        if before != timestamp:  # pragma: no cover
            raise storage_exceptions.IntegrityError(message="Inconsistent data. Retry.")

    # Cache control.
    _handle_cache_expires(request, bid, cid)

    data = {
        "metadata": metadata,
        "timestamp": timestamp,
        "changes": changes,
    }
    return data
